src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt

```diff
@@ -0,0 +1,145 @@
+        "islandcategory",
+        "^\\s+§r§b§lIsland$"
+    )
+    private val guestCategoryPattern by patternGroup.pattern(
+        "guestcategory",
+        "^\\s+§r§5§lGuests §r§f\\(\\d+\\)$"
+    )
+
+    @SubscribeEvent
+    fun onTabListUpdate(event: TabListUpdateEvent) {
+        if (!config.enabled) return
+        val onPrivateIslandGarden = IslandType.onPrivateWorld()
+        val guesting = IslandType.onPrivateWorld(guesting = true)
+        if (!(onPrivateIslandGarden || (config.onPublicIslands && !guesting) || (config.guestLeaveJoinMsgs && guesting))) return
+
+        val playersNew = mutableListOf<String>()
+
+        var inIslandCategory = false
+        val islandOwners = mutableListOf<String>()
+
+        val joinMessage = " §${if (config.leaveJoinColor) "a" else "e"}joined §ethe island."
+        val leaveMessage = " §${if (config.leaveJoinColor) "c" else "e"}left §ethe island."
+
+        for (line in event.tabList) {
+            if (guesting) {
+                islandCategoryPattern.matchMatcher(line) {
+                    inIslandCategory = true
+                }
+                if (inIslandCategory) offlinePlayerPattern.matchMatcher(line) {
+                    islandOwners.add(group("player"))
+                }
+                guestCategoryPattern.matchMatcher(line) {
+                    inIslandCategory = false
+                    for (player in islandOwners) {
+                        if (isPlayerKnown(player.removeColor())) {
+                            onKnownIsland = true
+                            return@matchMatcher
+                        }
+                    }
+                    onKnownIsland = false
+                    return@matchMatcher
+                }
+            }
+
+            cleanPlayerPattern.matchMatcher(line) {
+                val player = group("player")
+                playersNew.add(player)
+                if (guesting && inIslandCategory) { islandOwners.add(player) }
+                if (players.contains(player)) {
+                    return@matchMatcher
+                }
+                players.add(player)            // !onPrivateIslandGarden because a vanilla message gets sent
+                if (shouldSendMsg(player) && updatedSinceWorldSwitch && !onPrivateIslandGarden) {
+                    ChatUtils.chat(player + joinMessage)
+                }
+            }
+            rawPlayerPattern.matchMatcher(line) {
+                val player = group("player")
+                if (!playersNew.contains(player)) playersNew.add(player)
+            }
+        }
+
+        if (players.isEmpty()) return
+
+        if (players.size > 1) {
+            updatedSinceWorldSwitch = true
+        }
+
+        for ((index, player) in players.withIndex().reversed()) {
+            if (!playersNew.contains(player)) {
+                if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
+                players.removeAt(index)
```

Comment:
why do we reverse this list?
cant we simply use a filter, send the chat message for every result, then remove the list from the players list?

Reply:
It's reversed because otherwise, removing indexes would break it
https://discord.com/channels/997079228510117908/1094190239532208228/1190852439641952296

Reply:
Instead of reversing the list, you could create a duplication, iterate over it and remove the elements that way. No need to handle indexes at all then

Reply:
isn't it just 
```
val iter = player.reversed.iterator()
while(iter.hasNext()){
val player = iter.next()
if (!playersNew.contains(player)) {
                if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
                iter.remove()
}
```

Reply:
I would prefer:
```kt
for (player in players.filter { !playersNew.contains(it) }) {
    if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
    players.remove(player)
}

Reply:
> I would prefer:
> 
> ```kotlin
> for (player in players.filter { !playersNew.contains(it) }) {
>     if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
>     players.remove(player)
> }
> ```

Oh that's way better
I haven't used filter very much so I didn't think of that.